sqlglot/optimizer/predicate_pushdown.py

```diff
@@ -0,0 +1,71 @@
+from sqlglot import expressions as exp
+from sqlglot.optimizer.scope import traverse_scope
+
+
+def predicate_pushdown(expression):
+    """
+    Rewrite sqlglot AST to pushdown predicates in FROMS and JOINS
+
+    Example:
+        >>> import sqlglot
+        >>> sql = "SELECT * FROM (SELECT * FROM x AS x) AS y WHERE y.a = 1"
+        >>> expression = sqlglot.parse_one(sql)
+        >>> predicate_pushdown(expression).sql()
+        'SELECT * FROM (SELECT * FROM x AS x WHERE y.a = 1) AS y WHERE TRUE'
+
+    Args:
+        expression (sqlglot.Expression): expression to optimize
+    Returns:
+        sqlglot.Expression: optimized expression
+    """
+    for scope in reversed(traverse_scope(expression)):
+        where = scope.expression.args.get("where")
+
+        if not where or where.find(exp.Or):
```

Comment:
I imagine we're just starting with the basic case, but what about..

```sql
SELECT * 
FROM (...) AS x 
WHERE 
  x.a = 1 OR x.b = 2
```

And others.

What is the actual logical rule?

Reply:
right, this is more complex and requires converting to CNF (conjunctive normal form), i'm actually working on this right now but i have to figure out how to apply the distributive law